qiskit/synthesis/su4/siswap_decompose.py

```diff
@@ -0,0 +1,286 @@
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+# pylint: disable=invalid-name, non-ascii-name
+
+"""Synthesis of two-qubit unitaries using at most 3 applications of the sqrt(iSWAP) gate."""
+
+import cmath
+from typing import Optional, List, Union
+import numpy as np
+
+from qiskit.circuit.quantumcircuit import QuantumCircuit
+from qiskit.quantum_info.operators import Operator
+from qiskit.quantum_info.synthesis.two_qubit_decompose import TwoQubitWeylDecomposition
+from qiskit.circuit.library import SiSwapGate, XGate, YGate, ZGate
+from qiskit.synthesis.su4.utils import find_min_point, average_infidelity
+
+
+_EPS = 1e-12
+
+# the polytope accessible by two applications of the SiSwapGate
+# (red region in paper)
+_ID = np.array([0, 0, 0])
+_CNOT = np.array([np.pi / 4, 0, 0])
+_ISWAP = np.array([np.pi / 4, np.pi / 4, 0])
+_MID = np.array([np.pi / 4, np.pi / 8, np.pi / 8])
+_MIDDG = np.array([np.pi / 4, np.pi / 8, -np.pi / 8])
+_SISWAP = np.array([np.pi / 8, np.pi / 8, 0])
+_POLYTOPE = np.array([_ID, _CNOT, _ISWAP, _MID, _MIDDG])
+
+
+class SiSwapDecomposer:
+    """
+    A class for decomposing 2-qubit unitaries into at most 3 uses of the sqrt(iSWAP) gate.
+
+    This basis is attractive in that it is shorter/cheaper than iSWAP and generates a large
+    volume (~79%) of Haar-random SU(4) unitaries with only two applications, and an even
+    larger volume when approximation is enabled.
+
+    Args:
+        euler_basis (list(str)): single-qubit gates basis in the decomposition.
+
+    Reference:
+        1. C. Huang et al,
+           *Towards ultra-high fidelity quantum operations:
+           SQiSW gate as a native two-qubit gate (2021)*
+           `arXiv:2105.06074 <https://arxiv.org/abs/2105.06074>`_
+    """
+
+    def __init__(self, euler_basis: Optional[List[str]]):
+        # decomposer for the local single-qubit gates
+        from qiskit.transpiler.passes.optimization.optimize_1q_decomposition import (
+            Optimize1qGatesDecomposition,  # pylint: disable=cyclic-import
+        )
+
+        euler_basis = euler_basis or ["u"]
+        self._decomposer1q = Optimize1qGatesDecomposition(euler_basis)
+
+    def __call__(
+        self,
+        unitary: Union[Operator, np.ndarray],
+        basis_fidelity: Optional[float] = 1.0,
+        approximate: bool = True,
+        *,
+        _num_basis_uses: Optional[int] = None,
+    ) -> QuantumCircuit:
+        """Decompose a two-qubit unitary into using the sqrt(iSWAP) gate.
+
+        Args:
+            unitary: a 4x4 unitary to synthesize.
+            basis_fidelity: Fidelity of the iSWAP gate.
+            approximate: Approximates if basis fidelities are less than 1.0.
+            _num_basis_uses: force a particular approximation by passing a number in [0, 3].
+
+        Returns:
+            QuantumCircuit: Synthesized circuit.
+        """
+        if not approximate:
```

Comment:
That's a more general question: isn't `basis_fidelity` and `approximate` redundant?

Reply:
hmm good question. I was copying the structure of the XXDecomposer code. I can think about removing `approximate` but I have to see if there is some underlying reason. I'm making some changes to the whole qiskit.synthesis folder so I can have this on the list.

Reply:
`approximate` is used by `XXDecomposer` only when computing infidelities. The method used doesn't depend on the values of the fidelities in `basis_fidelity`, in particular does not depend on whether they are less than one. I don't know "less than one" means, maybe it's advice on what to choose if you know you are passing 1.0. The same holds if `basis_fidelity` is not a `dict`, but a float, as it is here. In this case infidelities are computed for several values of `strength` (which corresponds to a rotation angle I think) that are hardcoded in `XXDecomposer`. The infidelities are computed from the strengths by a linear model. If `approximate` is `True` the parameters of the linear model are determined from `basis_fidelity`. If `approximate` is `False` then `basis_fidelity` is ignored completely and infidelities are computed without any reference to parameters passed when calling the object nor those passed when constructing it. That's sort of the reverse of what I would expect, but that's what the code does.